full-service/src/service/transaction_builder.rs

```diff
@@ -430,20 +453,52 @@ impl<FPR: FogPubkeyResolver + 'static> WalletTransactionBuilder<FPR> {
-                &reserved_subaddresses,
-                &mut rng,
-            )?;
+            if let Some(spend_from_subaddress) = self.spend_only_from_subaddress.as_ref() {
+                // Send the change back to the subaddress that is spending the inputs.
+                // In the future, we may want to allow this to be a bit more configurable
+                let change_address = b58_decode_public_address(&spend_from_subaddress)?;
+                let reserved_subaddresses_for_spend_from_subaddress_mode =
+                    ReservedSubaddresses::from_subaddress_index(
+                        &account.account_key()?,
+                        self.spend_only_from_subaddress_index,
+                        None,
+                    );
+
+                // FIXME: We should change `transaction_builder.add_change_output` to accept
+                // a subaddress rather than the default change address. For now, the memo for
+                // transaction history will be incorrect on this change output, and it will
+                // instead validate only with the authenticated sender memo
+                let tx_out_context = transaction_builder.add_change_output(
+                    change_amount,
+                    &reserved_subaddresses_for_spend_from_subaddress_mode,
+                    &mut rng,
+                )?;
 
-            let change_txo = OutputTxo {
-                tx_out: tx_out_context.tx_out,
-                recipient_public_address: reserved_subaddresses.change_subaddress.clone(),
-                confirmation_number: tx_out_context.confirmation,
-                amount: change_amount,
-                shared_secret: Some(tx_out_context.shared_secret),
-            };
-            change_txos.push(change_txo);
+                let change_txo = OutputTxo {
+                    tx_out: tx_out_context.tx_out,
+                    recipient_public_address: change_address,
+                    confirmation_number: tx_out_context.confirmation,
+                    amount: change_amount,
+                    shared_secret: Some(tx_out_context.shared_secret),
+                };
+                change_txos.push(change_txo);
+            } else {
+                // Send the change to the reserved change subaddress for the account
+                let tx_out_context = transaction_builder.add_change_output(
+                    change_amount,
+                    &reserved_subaddresses,
+                    &mut rng,
+                )?;
+
+                let change_txo = OutputTxo {
+                    tx_out: tx_out_context.tx_out,
+                    recipient_public_address: reserved_subaddresses.change_subaddress.clone(),
+                    confirmation_number: tx_out_context.confirmation,
+                    amount: change_amount,
+                    shared_secret: Some(tx_out_context.shared_secret),
+                };
+                change_txos.push(change_txo);
+            }
```

Comment:
consider doing only the differences in different blocks
```suggestion
            let (change_address, change_destination) = match self.spend_only_from_subaddress {
                Some(ref subaddress) => {
                    let change_address = b58_decode_public_address(subaddress)?;
                    let reserved_subaddresses = ReservedSubaddresses::from_subaddress_index(
                        &account.account_key()?,
                        self.spend_only_from_subaddress_index,
                        None,
                    );
                    (change_address, reserved_subaddresses)
                }
                None => (
                    reserved_subaddresses.change_subaddress.clone(),
                    reserved_subaddresses.clone(),
                ),
            };
            let tx_out_context = transaction_builder.add_change_output(
                change_amount,
                &reserved_subaddresses,
                &mut rng,
            )?;

            let change_txo = OutputTxo {
                tx_out: tx_out_context.tx_out,
                recipient_public_address: change_address,
                confirmation_number: tx_out_context.confirmation,
                amount: change_amount,
                shared_secret: Some(tx_out_context.shared_secret),
            };
            change_txos.push(change_txo);
```

TBH the `let (..) match` logic should probably be moved outside of the for loop, this would alleviate one of the clones as well as thin up the logic that's critical to the for loop

I'm also under the impression the FIXME no longer applies with the latest change of using `ReservedSubaddresses::from_subaddress_index()`